axlearn/common/pipeline.py

```diff
@@ -305,80 +310,99 @@ def vmap_fn(
-                    )
-
-                # Per-timestep inputs.
-                # Each leaf tensor in `prng_key_t` and `x_t` has shape [N, ...].
-                prng_key_t, input_t, x_t = scan_t
-                carry_input_t = jax.tree_util.tree_map(
-                    compute_carry_input, input_t, carry_output_t_1, carry_partition_spec
+                # Input state.
+                t = carry_in["t"]
+                carry_output_t_1 = carry_in["carry_output_t_1"]
+                per_stage_inputs = carry_in["per_stage_inputs"]
+
+                # Per-timestep inputs. Each leaf tensor has shape [N, ...] or [1, ...].
+                prng_key_t, x_t = xs_t
+
+                # Compute vmap inputs. When t >= m, we feed dummy inputs to the pipeline until the
+                # pipeline is flushed. Note that at the end of all iterations we only extract the
+                # last-stage outputs from the stacked vmap outputs.
+                # Leaves are of shape [N, ...] representing per-stage inputs.
+                vmap_in = self._compute_carry_input(per_stage_inputs, carry_output_t_1, t=t)
+
+                # Use stop_gradient for invalid (bubble) microbatch iterations. This jnp.where will
+                # be optimized away by XLA, but in the backward pass it will be masking with zeros.
+                state = jax.tree_util.tree_map(
+                    lambda x: jnp.where(self._is_valid_stage(x, t=t), x, jax.lax.stop_gradient(x)),
+                    layer_context.state,
                 )
 
                 # Parallel processing along the N axis.
-                vmap_out = jax.vmap(vmap_fn)(layer_context.state, prng_key_t, carry_input_t, x_t)
+                vmap_out = jax.vmap(vmap_fn)(state, prng_key_t, vmap_in, x_t)
                 self.vlog(3, "vmap_out.output_collection=%s", shapes(vmap_out["output_collection"]))
-                return vmap_out["carry"], vmap_out
 
-            carry_t0 = jax.tree_util.tree_map(
-                lambda x: jnp.tile(jnp.zeros_like(x[:1]), [n] + [1] * (x.ndim - 1)), carry
+                # Output state.
+                carry_out = dict(
+                    t=t + 1,
+                    carry_output_t_1=vmap_out["carry"],
+                    per_stage_inputs=per_stage_inputs,
+                )
+                # TODO(markblee): Consider slicing out just the last-stage outputs of vmap_out.
+                # Note that vmap outputs are typically sharded over stages and may incur extra
+                # communication per-iteration (e.g. from broadcasting last stage outputs).
+                return carry_out, vmap_out
+
+            state_t0 = dict(
+                # Current loop iteration.
+                t=jnp.array(0, dtype=jnp.int32),
+                # [N, microbatch_size, ...].
+                carry_output_t_1=jax.tree_util.tree_map(
+                    lambda x: jnp.zeros((n,) + x.shape[1:]), carry
```

Comment:
see this PR: https://github.com/markblee/axlearn/pull/1

Reply:
Thanks -- how can I reproduce this error? (Also added `x.dtype`, in case that resolves it.)

Reply:
Do we have a test for bfloat16?

Reply:
Added a test that fails without the change.